### ISSUE TITLE
[no-assignment-of-untracked-properties-used-in-tracking-contexts] False positive with constructors?

### ISSUE THREAD

**Issue by boris-petrov** [ember-cli/eslint-plugin-ember#867]
Not absolutely sure about that. I'm getting that warning for some assignments in a constructor. That is, I have a native JS class that uses `@computed` in it and sets in the constructor some properties that are used in that `@computed`... should this assignment be done via `Ember.set` or is this a false-positive?

**Comment by bmish** [comment]
I am assuming you are talking about this example:

```js
import Component from '@ember/component';
import { computed } from '@ember/object';

export default class MyComponent extends Component {
  constructor() {
    super();

    this.x = 123; // this should use `set` according to ember/no-assignment-of-untracked-properties-used-in-tracking-contexts
  }

  @computed('x')
  get myProp() {
    return this.x;
  }
}
```

Using `set` in the constructor seems like a good thing to do for consistency, given that we should be using `set` when modifying this property anywhere else in the file.

But is it actually required to use `set` for this property in the constructor? I don't believe so. So I'd be open to any of these changes:

1. Update the rule to always ignore constructors.
2. Add an `ignoreConstructors` option (default true).
3. Add an `ignoreConstructors` option (default false).

@rwjblue @pzuraq @mongoose700 thoughts?

**Comment by mongoose700** [comment]
There can be cases where setting things without `set` happens to work, but I don't think anyone should ever rely on it. I think it's possible for it to fail in a constructor as well:
```
import Component from '@ember/component';
import { computed } from '@ember/object';

export default class MyComponent extends Component {
  constructor() {
    super();

    this.y = this.myProp;
    this.x = 123; // I believe this will fail because `x` is now being observed
  }

  @computed('x')
  get myProp() {
    return this.x;
  }
}
```

**Comment by boris-petrov** [comment]
@bmish - my case is even simpler - I don't even extend from `Component` - as I said, it's a plain JS class:

```typescript
import { computed } from '@ember/object';

export default class Foo {
  constructor() {
    this.x = 123; // this should use `set` according to ember/no-assignment-of-untracked-properties-used-in-tracking-contexts
  }

  @computed('x')
  get myProp() {
    return this.x;
  }
```

Not that it matters.

I'm voting for option 1 even though @mongoose700 has a point. But his example is horrible code anyways and nobody should be writing it. :smiley:

**Comment by bmish** [comment]
My understanding is that `set` needs to be used with any computed property dependent keys, regardless of what type of class is involved. And based on @mongoose700's comment, it seems that it is correct to use `set` in this situation, even though the code may work without it much of the time. So I'm currently leaning toward not making any change to the rule...

**Comment by rwjblue** [comment]
> My understanding is that set needs to be used with any computed property dependent keys, regardless of what type of class is involved.

This is correct, **but** `constructor` runs _before_ any tracking is setup and therefore `Ember.set` isn't "required".

**Comment by bmish** [comment]
@rwjblue thanks for the confirmation! In that case, I can update the rule to always ignore any assignments in `constructor`.

**Comment by rwjblue** [comment]
I personally think its fine to have the rule suggest using `set` here making it much simpler to reason about ("always use `set`")

**Comment by rwjblue** [comment]
Hehe, sorry for being unclear. I think its super fine to go either way, I don't feel strongly. One the one hand: consistency; on the other hand: doing "extra work" in the constructor...

**Comment by bmish** [comment]
Hmm! I'm a big fan of the consistency argument. It's confusing to use `set` in some places but assignment in other places for the same property. So perhaps enforcing consistency even if `set` isn't needed in the constructor is actually a feature and not a bug.
